File: java/src/main/java/net/asam/openscenario/parser/type/XmlComplexTypeParser.java
/*
 * Copyright 2020 RA Consulting
 *
 * RA Consulting GmbH licenses this file under the Apache License,
 * Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package net.asam.openscenario.parser.type;

import java.util.List;
import java.util.Map;

import org.w3c.dom.Element;
import org.w3c.dom.NamedNodeMap;
import org.w3c.dom.Node;

import net.asam.openscenario.common.ErrorLevel;
import net.asam.openscenario.common.FileContentMessage;
import net.asam.openscenario.common.IParserMessageLogger;
import net.asam.openscenario.common.Textmarker;
import net.asam.openscenario.impl.BaseImpl;
import net.asam.openscenario.parser.ParserContext;
import net.asam.openscenario.parser.XmlModelGroupParser;
import net.asam.openscenario.parser.XmlParserBase;
import net.asam.openscenario.simple.struct.IndexedElement;
import net.asam.xml.indexer.Position;

/**
 * Parsing an XSD:complexType
 *
 * @author Andreas Hege - RA Consulting
 */
public abstract class XmlComplexTypeParser<T extends BaseImpl> extends XmlParserBase<T>
    implements IXmlTypeParser<T> {

  protected XmlModelGroupParser<T> subElementParser;
  private Map<String, IAttributeParser<T>> attributeNameToAttributeParser;

  /**
   * A dictionary that maps an attribute names to their parser.
   *
   * @return the name to parser dictionary
   */
  protected abstract Map<String, IAttributeParser<T>> getAttributeNameToAttributeParserMap();

  /**
   * Constructor
   *
   * @param messageLogger to log messages during parsing process
   * @param filename of the file the parser is operating on.
   */
  public XmlComplexTypeParser(IParserMessageLogger messageLogger, String filename) {
    super(messageLogger, filename);
  }
  /**
   * parses all atrributes and fills them into the object
   *
   * @param indexedElement the element to be parsed
   * @param object the object that will be filled during the parsing process.
   */
  protected void parseAttributes(IndexedElement indexedElement, T object) {
    attributeNameToAttributeParser = getAttributeNameToAttributeParserMap();
    Element element = indexedElement.getElement();
    NamedNodeMap attributeMap = element.getAttributes();
    Position position = indexedElement.getStartElementLocation();
    for (int i = 0; i < attributeMap.getLength(); i++) {
      Node attributeNode = attributeMap.item(i);

      String attributeName = attributeNode.getNodeName();
      String attributeValue = attributeNode.getNodeValue();
      IAttributeParser<T> parser = attributeNameToAttributeParser.get(attributeName);
      Position attributeStartPosition = indexedElement.getAttributeStartPosition(attributeName);
      Position attributeEndPosition = indexedElement.getAttributeEndPosition(attributeName);

      if (attributeName.contains(":")) {
        messageLogger.logMessage(
            new FileContentMessage(
                "Ignoring attribute '" + attributeName + "'",
                ErrorLevel.INFO,
                new Textmarker(
                    attributeStartPosition.getLine(),
                    attributeStartPosition.getColumn(),
                    filename)));
      } else if (parser == null) {
        messageLogger.logMessage(
            new FileContentMessage(
                "Unknown attribute '" + attributeName + "'",
                ErrorLevel.ERROR,
                new Textmarker(
                    attributeStartPosition.getLine(),
                    attributeStartPosition.getColumn(),
                    filename)));
      } else {
        parser.parse(
            attributeStartPosition, attributeEndPosition, attributeName, attributeValue, object);
        // Remove
        attributeNameToAttributeParser.remove(attributeName);
      }
    }

    for (String attributeName : attributeNameToAttributeParser.keySet()) {
      if (attributeNameToAttributeParser.get(attributeName).getMinOccur() > 0) {
        messageLogger.logMessage(
            new FileContentMessage(
                "Required attribute '" + attributeName + "' is missing ",
                ErrorLevel.ERROR,
                new Textmarker(position.getLine(), position.getColumn(), filename)));
      }
    }
  }

  @Override
  public void parseElement(IndexedElement indexedElement, ParserContext parserContext, T object) {
    // The element was requested before, so it cannot be null
    Position position = indexedElement.getStartElementLocation();
    String elementName = indexedElement.getElement().getNodeName();
    if (indexedElement.isMixedNode()) {
      messageLogger.logMessage(
          new FileContentMessage(
              "Element '" + elementName + "' contains text. Only subelements allowed.",
              ErrorLevel.ERROR,
              new Textmarker(position.getLine(), position.getColumn(), filename)));
    }
    parseAttributes(indexedElement, object);
    parseSubElements(indexedElement.getSubElements(), parserContext, object);
    parserContext.setLastElementParsed(indexedElement);
    Position startPosition = indexedElement.getStartElementLocation();
    object.setStartMarker(
        new Textmarker(startPosition.getLine(), startPosition.getColumn(), filename));
    Position endPosition = indexedElement.getStartElementLocation();
    object.setEndMarker(new Textmarker(endPosition.getLine(), endPosition.getColumn(), filename));
  }

  /**
   * Checks for paramterized value ($-notation)
   *
   * @param value value
   * @return tue if the attribute has a prameter instead of a literal value
   */
  protected boolean isParametrized(String value) {
    // Only Dollar will result in "$"
    return value.startsWith("$") && value.length() > 1;
  }
  /**
   * Stripes the '$' from a name when the value starts with '$'
   *
   * @param value value that might start with a '$'
   * @return value without a starting '$'
   */
  protected String stripDollarSign(String value) {
    if (value.startsWith("$")) {
      if (value.length() > 1) {
        return value.substring(1);
      } else {
        return "";
      }
    } else {
      return value;
    }
  }

  @Override
  public void parseSubElements(
      List<IndexedElement> indexedElements, ParserContext parserContext, T object) {

    subElementParser.parseSubElements(indexedElements, parserContext, object);
  }
}

File: cpp/openScenarioLib/src/parser/type/XmlComplexTypeParser.cpp
/*
 * Copyright 2020 RA Consulting
 *
 * RA Consulting GmbH licenses this file under the Apache License,
 * Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "XmlComplexTypeParser.h"
#include <regex>

namespace NET_ASAM_OPENSCENARIO
{
    std::map<std::string, std::shared_ptr<IAttributeParser>> XmlComplexTypeParser::GetAttributeNameToAttributeParserMap()
    {
        return {};
    }


	XmlComplexTypeParser::XmlComplexTypeParser(IParserMessageLogger& messageLogger, std::string& filename): XmlParserBase(messageLogger, filename) {}
	XmlComplexTypeParser::~XmlComplexTypeParser() = default;
	
    void XmlComplexTypeParser::ParseAttributes(std::shared_ptr<IndexedElement>& indexedElement, std::shared_ptr <BaseImpl> object)
    {
        _attributeNameToAttributeParser = GetAttributeNameToAttributeParserMap();
        const auto kElement = indexedElement->GetElement();
        std::vector<const tinyxml2::XMLAttribute *> attributeMap;

        if (kElement->FirstAttribute() != nullptr)
        {
            attributeMap.push_back(kElement->FirstAttribute());
            while (attributeMap[attributeMap.size() - 1]->Next() != nullptr)
            {
                attributeMap.push_back(attributeMap[attributeMap.size() - 1]->Next());
            }
        }
        auto position = indexedElement->GetStartElementLocation();
        for (auto&& attribute: attributeMap)
        {
            std::string attributeName = attribute->Name();
            std::string attributeValue = attribute->Value();
            std::shared_ptr<IAttributeParser> parser;
            const auto kIt = _attributeNameToAttributeParser.find(attributeName);
            if (kIt != _attributeNameToAttributeParser.end())
                parser = kIt->second;

            auto attributeStartPosition = indexedElement->GetAttributeStartPosition(attributeName);
			auto attributeEndPosition = indexedElement->GetAttributeEndPosition(attributeName);
			auto attributeStartValuePosition = indexedElement->GetAttributeStartValuePosition(attributeName);

            if (attributeName.find(":") != std::string::npos)
            {
                auto msg = FileContentMessage("Ignoring attribute '" + attributeName + "'", INFO, Textmarker(attributeStartPosition.GetLine(), attributeStartPosition.GetColumn(), this->_filename));
                this->_messageLogger.LogMessage(msg);
            }
            else if (kIt == _attributeNameToAttributeParser.end())
            {
                auto msg = FileContentMessage("Unknown attribute '" + attributeName + "'", ERROR, Textmarker(attributeStartPosition.GetLine(), attributeStartPosition.GetColumn(), this->_filename));
                this->_messageLogger.LogMessage(msg);
            }
            else
            {
                parser->Parse(attributeStartPosition, attributeEndPosition, attributeStartValuePosition, attributeName, attributeValue, object);
                // Remove 
               _attributeNameToAttributeParser.erase(attributeName);
            }
        }

        for (auto&& elem : _attributeNameToAttributeParser) 
        {
            if (elem.second->GetMinOccur() > 0)
            {
                auto msg = FileContentMessage("Required attribute '" + elem.first + "' is missing ", ERROR, Textmarker(position.GetLine(), position.GetColumn(), this->_filename));
                this->_messageLogger.LogMessage(msg);
            }
        }
    }

    void XmlComplexTypeParser::ParseElement(std::shared_ptr<IndexedElement>& indexedElement, std::shared_ptr <ParserContext>& parserContext, std::shared_ptr <BaseImpl> object)
    {
        // The element was requested before, so it cannot be null
        const auto kPosition = indexedElement->GetStartElementLocation();
        const std::string kElementName = indexedElement->GetElement()->Name();
        if (indexedElement->IsMixedNode())
        {
            auto msg = FileContentMessage("Element '" + kElementName + "' contains text. Only subelements allowed.", ERROR, Textmarker(kPosition.GetLine(), kPosition.GetColumn(), this->_filename));
            this->_messageLogger.LogMessage(msg);
        }
        ParseAttributes(indexedElement, object);
        auto subElements = indexedElement->GetSubElements();
        ParseSubElements(subElements, parserContext, object);
        parserContext->SetLastElementParsed(indexedElement);
        const auto kStartPosition = indexedElement->GetStartElementLocation();
        object->SetStartMarker(Textmarker(kStartPosition.GetLine(), kStartPosition.GetColumn(), this->_filename));
        const auto kEndPosition = indexedElement->GetEndElementLocation();
        object->SetEndMarker(Textmarker(kEndPosition.GetLine(), kEndPosition.GetColumn(), this->_filename));
    }
	
    bool XmlComplexTypeParser::IsParametrized(std::string& value)
    {
        // Only Dollar will result in "$"
        return value[0] == '$' && value.length() > 1;
    }

	bool XmlComplexTypeParser::IsExpression(std::string& value)
	{
		// Only Dollar will result in "$"
		std::smatch base_match;
		const std::regex base_regex("^\\s*\\$\\s*\\{");
		return std::regex_match(value,base_match, base_regex);
	}
	
    /**
     * Stripes the '$' from a name when the value starts with '$'
     * @param value value that might start with a '$'
     * @return value without a starting '$'
     */
    std::string XmlComplexTypeParser::StripDollarSign(std::string& value)
    {
        if (value[0] == '$')
        {
            if (value.length() > 1)
            {
                return value.substr(1);
            }
            else
            {
                return "";
            }
        }

        return value;
    }


    void XmlComplexTypeParser::ParseSubElements(std::vector<std::shared_ptr<IndexedElement>>& indexedElements, std::shared_ptr<ParserContext>& parserContext, std::shared_ptr<BaseImpl> object)
    {
        _subElementParser->ParseSubElements(indexedElements, parserContext, object);
    }

}
